test/fail_compilation/test19961.d

```diff
@@ -0,0 +1,46 @@
+/*
+REQUIRED_ARGS:
+PERMUTE_ARGS:
+TEST_OUTPUT:
+---
+fail_compilation/test19961.d(20): Error: `immutable` delegate `test15306.main.__dgliteral1` cannot access mutable data `i`
+fail_compilation/test19961.d(28): Error: `const` delegate test15306.main.__dgliteral2 cannot access mutable data `i`
+fail_compilation/test19961.d(35): Error: `inout` delegate test15306.main.__dgliteral2 cannot access mutable data `i`
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=15306
+// https://issues.dlang.org/show_bug.cgi?id=19961
+void main()
+{
+    // immutable cannot access mutable
+    int i = 42;
+    auto dg1 = delegate void() immutable
+    {
+        i++;
+        /*auto inner = i;*/ // incorectly gives error
+    };
+
+    // const cannot access mutable
+    int* p = &i;
+    auto dg2 = delegate int() const
+    {
+        i++;          // successfully rejected
+        //int j = *p; // incorrectly rejected
```

Comment:
This needs to be fixed before the PR can be merged. I believe that if you are able to fix the root cause and change the type of the `p` inside the delegate it automatically fix this as well ;)

Reply:
The problem is that the context pointer's qualifiers are not applied at all to access from within the nested function. I'm probably going to end up deleting the entire loop and else branch, the comment of the cascaded nested function is a complete lie, `i` can modify `fx` even if it is `pure`.

Reply:
Isn't it simple the case that the context pointer is not qualified? It mustn't be, or the type checker would be freaking out everywhere. Wherever the function receives the `Context*` argument, if it swaps for `qualified(Context)*`, then everything should wring out naturally?